src/main/java/org/wildfly/security/authz/jacc/ElytronPolicyConfiguration.java

```diff
@@ -0,0 +1,231 @@
+        checkIfInOpenState();
+
+        this.rolePermissions.computeIfAbsent(roleName, s -> new Permissions()).add(permission);
+    }
+
+    @Override
+    public void addToRole(String roleName, PermissionCollection permissions) throws PolicyContextException {
+        checkNotNullParam("roleName", roleName);
+        checkNotNullParam("permissions", permissions);
+
+        Enumeration<Permission> elements = permissions.elements();
+
+        while (elements.hasMoreElements()) {
+            addToRole(roleName, elements.nextElement());
+        }
+    }
+
+    @Override
+    public void addToUncheckedPolicy(Permission permission) throws PolicyContextException {
+        checkNotNullParam("permission", permission);
+        checkIfInOpenState();
+
+        this.uncheckedPermissions.add(permission);
+    }
+
+    @Override
+    public void addToUncheckedPolicy(PermissionCollection permissions) throws PolicyContextException {
+        checkNotNullParam("permissions", permissions);
+
+        Enumeration<Permission> elements = permissions.elements();
+
+        while (elements.hasMoreElements()) {
+            addToUncheckedPolicy(elements.nextElement());
+        }
+    }
+
+    @Override
+    public void commit() throws PolicyContextException {
+        synchronized (this) {
+            if (isDeleted()) {
+                throw log.authzInvalidStateForOperation(this.state.name());
+            }
+
+            transitionTo(State.IN_SERVICE);
+        }
+    }
+
+    @Override
+    public void delete() throws PolicyContextException {
+        synchronized (this) {
+            transitionTo(State.DELETED);
+            this.uncheckedPermissions = new Permissions();
+            this.excludedPermissions = new Permissions();
+            this.rolePermissions.clear();
+            this.linkedPolicies.remove(this);
+        }
+    }
+
+    @Override
+    public String getContextID() throws PolicyContextException {
+        return this.contextId;
+    }
+
+    @Override
+    public boolean inService() {
+        return State.IN_SERVICE.equals(this.state);
+    }
+
+    @Override
+    public void linkConfiguration(PolicyConfiguration link) throws PolicyContextException {
+        checkNotNullParam("link", link);
+        checkIfInOpenState();
+
+        synchronized (this.linkedPolicies) {
+            if (getContextID().equals(link.getContextID())) {
+                throw log.authzLinkSamePolicyConfiguration(getContextID());
+            }
+
+            this.linkedPolicies.add(this);
+
+            if (!this.linkedPolicies.add(link)) {
+                return;
+            }
+
+            ElytronPolicyConfiguration linkedPolicyConfiguration = (ElytronPolicyConfiguration) link;
+
+            linkedPolicyConfiguration.linkConfiguration(this);
+            // policies share the same set of linked policies, so we can remove policies from the set when they are deleted.
+            this.linkedPolicies = linkedPolicyConfiguration.getLinkedPolicies();
```

Comment:
Sorry, missed that issue. I've also removed the volatile and now it is using sync blocks.